File: chapter3.hs
{-# LANGUAGE DeriveDataTypeable #-}

module  Chapter3 (Chapter3(..)) where
  import Prelude hiding (tail, head)
  import Data.Typeable
  import Data.List (reverse)
  import Control.Exception

  data Chapter3 = Chapter3
  instance Exception HeapException

  data HeapException = HeapEmpty
                     deriving (Show, Typeable)

  class Emptiable h where
    empty :: Ord a  => h a
    isEmpty :: Ord a => h a -> Bool

  class Insertable h where
    insert :: Ord a => a -> h a -> h a

  class (Emptiable s, Insertable s) => Set s where
    member :: Ord a => a -> s a -> Bool

  class (Emptiable h, Insertable h) => Heap h where
    merge :: Ord a => h a -> h a -> h a
    findMin :: Ord a => h a -> a
    deleteMin :: Ord a => h a -> h a

  data LeftistHeap a = E 
                     | LeftistHeap Int a (LeftistHeap a) (LeftistHeap a)
                     deriving (Show)
    
  instance Emptiable LeftistHeap where
    empty = E
    isEmpty E = True
    isEmpty _ = False

  rank E = 0
  rank (LeftistHeap r _ _ _) = r
  makeT x a b
    | rank a > rank b = LeftistHeap (rank b + 1) x a b
    | otherwise = LeftistHeap (rank a + 1) x b a
  
  instance Heap LeftistHeap where
    merge h1 E = h1
    merge E h2 = h2
    merge h1@(LeftistHeap _ x a1 b1) h2@(LeftistHeap _ y a2 b2)
      | x <= y = makeT x a1 (merge b1 h2)
      | otherwise = makeT y a2 (merge h1 b2)
    findMin E = throw HeapEmpty
    findMin (LeftistHeap _ x h1 h2) = x
    deleteMin (LeftistHeap _ x h1 h2) = merge h1 h2
    deleteMin _ = E

  instance Insertable LeftistHeap where
    insert x h = merge (LeftistHeap 1 x E E) h


  data T a = T Int a [T a]

  newtype BinomialHeap a = BinomialHeap [T a]
  
  instance Emptiable BinomialHeap where
    empty = BinomialHeap []
    isEmpty (BinomialHeap []) = True
    isEmpty _ = False

  rank' (T r x c) = r
  root' (T r x c) = x
  
  link t1@(T r1 x1 c1) t2@(T _ x2 c2) 
    | x1 <= x2 = T (r1 + 1) x1 (t2:c1)
    | otherwise = T (r1 + 1) x2 (t1:c2)

  insTree t [] = [t]
  insTree t ts@(t':ts') 
    | rank' t < rank' t' =  t:ts
    | otherwise = insTree (link t t') ts'
  
  instance Insertable BinomialHeap where
    insert x (BinomialHeap ts) = BinomialHeap (insTree (T 0 x []) ts)

  merge' ts1 [] = ts1
  merge' [] ts2 = ts2
  merge' ts1@(t1:ts1') ts2@(t2:ts2') =
          if rank' t1 < rank' t2 then t1:(merge' ts1' ts2)
          else if rank' t2 < rank' t1 then t2:(merge' ts1 ts2')
          else insTree (link t1 t2) (merge' ts1' ts2')

  removeMinTree [] = throw HeapEmpty
  removeMinTree [t] = (t, [])
  removeMinTree (t:ts)
    | root' t <= root' t' = (t, ts)
    | otherwise = (t', t:ts')
    where 
       (t', ts') = removeMinTree ts

  instance Heap BinomialHeap where
    merge (BinomialHeap ts1) (BinomialHeap ts2) = BinomialHeap (merge' ts1 ts2)
    findMin (BinomialHeap ts) = root' t
      where (t, _) = removeMinTree ts
    deleteMin (BinomialHeap ts) = BinomialHeap (merge' (reverse ts1) ts2)
      where (T _ x ts1, ts2) = removeMinTree ts


  data Color = R | B
  data RB a = E'
            | RB Color (RB a) a (RB a)

  instance Emptiable RB where
    empty = E'
    isEmpty E' = True
    isEmpty _ = False
  
  balance t = case t of 
    (RB B (RB R (RB R a x b) y c) z d) -> balanced a x b y c z d
    (RB B (RB R a x (RB R b y c)) z d) -> balanced a x b y c z d
    (RB B a x (RB R (RB R b y c) z d)) -> balanced a x b y c z d
    (RB B a x (RB R b y (RB R c z d))) -> balanced a x b y c z d
    (RB a b c d) -> (RB a b c d)
    where 
      balanced a x b y c z d = RB R (RB B a x b) y (RB B c z d)

  instance Insertable RB where
    insert x t = 
      let ins t = case t of
            E' -> RB R E' x E'
            (RB color a y b) -> if x < y then balance (RB color (ins a) y b)
                                else if y < x then balance (RB color a y (ins b))
                                else t
      in
         case ins t of
          (RB _ a y b) -> RB B a y b
          _ -> error "Impossible pattern for insert"

  instance Set RB where
    member x E' = False
    member x (RB _ a y b)
      | x < y = member x a
      | y < x = member x b
      | otherwise = True

File: chapter3.ml
exception Empty
exception Impossible_pattern of string

module type ORDERED = sig
  type t

  val eq : t -> t -> bool
  val lt : t -> t -> bool
  val leq : t -> t -> bool
end
    
module type SET = sig
  type elem
  type set

  val empty : set
  val insert : elem -> set -> set
  val member : elem -> set -> bool
end

module type Heap = sig
  type elem
  type heap

  val empty : heap
  val isEmpty : heap -> bool
  val insert : elem -> heap -> heap
  val merge : heap -> heap -> heap
  val findMin : heap -> elem
  val deleteMin : heap -> heap
end

 
module LeftistHeap (Element : ORDERED) : (Heap with type elem = Element.t) =
struct
  type elem = Element.t
  type tree = E | T of int * elem * tree * tree
  type heap = tree

  let empty = E
  let isEmpty = function
    | E -> true 
    | _ -> false
  let rank = function
    | E -> 0
    | T(r, _, _, _) -> r

  let makeT x a b = 
    if rank a > rank b then T(rank b + 1, x, a, b)
    else T(rank a + 1, x, b, a)
     
  let rec merge h1 h2 = 
    match h1, h2 with
      | _, E -> h1
      | E, _ -> h2
      | T(_, x, a1, b1), T(_, y, a2, b2) -> 
          if x <= y then makeT x a1 (merge b1 h2)
          else makeT y a2 (merge h1 b2)

  let insert x h = merge (T(1, x, E, E)) h
  let findMin = function
    | T(_, x, _, _) -> x
    | E -> raise Empty
  let deleteMin = function
    | T(_, x, a, b) -> merge a b
    | _ -> E
end

module BinomialHeap (Element : ORDERED) : (Heap with type elem = Element.t) = 
struct
  type elem = Element.t
  type tree = T of int * elem * tree list
  type heap = tree list

  let empty = []
  let isEmpty = function
    | [] -> false 
    | _ -> true
  let rank (T(r, x, c)) = r
  let root (T(r, x, c)) = x
  let link  (T(r1, x1, c1) as t1) (T(_, x2, c2) as t2) =
      if Element.leq x1 x2 then T(r1 + 1, x1, t2::c1)
      else T(r1 + 1, x2, t1::c2)

  let rec insTree t = function
    | [] -> [t]
    | t'::ts' as ts -> if rank t < rank t' then t :: ts
      else insTree (link t t') ts'

  let insert x ts = insTree (T(0, x, [])) ts

  let rec merge ts1 ts2 = 
    match ts1, ts2 with
      | _, [] -> ts1
      | [], _ -> ts2
      | t1 :: ts1', t2 :: ts2' -> 
          if rank t1 < rank t2 then t1::merge ts1' ts2
          else if rank t2 < rank t1 then t2::merge ts1 ts2'
          else insTree (link t1 t2) (merge ts1' ts2')

  let rec removeMinTree = function
    | [] -> raise Empty
    | [t] -> (t, [])
    | (t::ts) -> let (t', ts') = removeMinTree ts
      in if Element.leq (root t) (root t') then (t, ts) else (t', t::ts')

  let findMin ts = let (t, _) = removeMinTree ts in root t
  let deleteMin ts = let (T(_, x, ts1), ts2) = removeMinTree ts in
    merge (List.rev ts1) ts2
end

                                                      
module RedBlackSet (Element : ORDERED) : (SET with type elem = Element.t) = 
struct
  type elem = Element.t
  type color = R | B
  type tree = E | T of color * tree * elem * tree
  type set = tree

  let empty = E

  let rec member x = function
    | E -> false
    | T(_, a, y, b) ->
        if Element.lt x y then member x a
        else if Element.lt y x then member x b
        else true

  let balance = function
    | B, T (R, T (R, a, x, b), y, c), z, d
    | B, T (R, a, x, T (R, b, y, c)), z, d
    | B, a, x, T (R, T (R, b, y, c), z, d)
    | B, a, x, T (R, b, y, T (R, c, z, d)) ->
        T (R, T (B, a, x, b), y, T (B, c, z, d))
    | a, b, c, d -> T (a, b, c, d)

  let insert x s =
    let rec ins = function
      | E -> T (R, E, x, E)
      | T (color, a, y, b) as s ->
          if Element.lt x y then balance (color, ins a, y, b)
          else if Element.lt y x then balance (color, a, y, ins b)
          else s
    in
      match ins s with
        | T(_, a, y, b) -> T (B, a, y, b)
        | _ -> raise (Impossible_pattern "insert")
end
